afero.Rename errors when copying mongod binary from /tmp mounted on different filesystem

My system has `/tmp` mounted on a different filesystem, so when `afs.Rename` is called to copy the mongod binary from `/tmp` to the cache, I get the following error:

```
panic: error writing mongod binary from /tmp/192458888 to /home/devin/.cache/memongo/mongodb-linux-x86_64-debian10-4_2_16_tgz_ffac481192/mongod: rename /tmp/192458888 /home/devin/.cache/memongo/mongodb-linux-x86_64-debian10-4_2_16_tgz_ffac481192/mongod: invalid cross-device link
```

This is because it is likely calling `os.Rename`, which cannot move files across filesystems. This can be worked around by [copying the file and then deleting the original if the copy was successful](https://gist.github.com/var23rav/23ae5d0d4d830aff886c3c970b8f6c6b).

--- a/mongobin/getOrDownload.go
+++ b/mongobin/getOrDownload.go
@@ -5,10 +5,12 @@ import (
 	"compress/gzip"
 	"crypto/sha256"
 	"encoding/hex"
+	"errors"
 	"fmt"
 	"io"
 	"net/http"
 	"net/url"
+	"os"
 	"path"
 	"regexp"
 	"strings"
@@ -130,6 +132,22 @@ func GetOrDownloadMongod(urlStr string, cachePath string, logger *memongolog.Log
 
 	renameErr := Afs.Rename(mongodTmpFile.Name(), mongodPath)
 	if renameErr != nil {
+		linkErr := &os.LinkError{}
+		if errors.As(renameErr, &linkErr) {
+			// If /tmp is on another filesystem, we have to copy the file instead.
+			logger.Debugf("Unable to move %s to %s, copying instead", mongodTmpFile.Name(), mongodPath)
+			mongodFile, err := Afs.Create(mongodPath)
+			if err != nil {
+				return "", fmt.Errorf("creating mongod binary at %s: %s", mongodTmpFile, err)
+			}
+			defer mongodFile.Close()
+
+			_, copyErr := io.Copy(mongodFile, mongodTmpFile)
+			if copyErr != nil {
+				fmt.Errorf("error copying mongod binary from %s to %s: %s", mongodTmpFile.Name(), mongodPath, copyErr)
+			}
+		}
+
 		return "", fmt.Errorf("error writing mongod binary from %s to %s: %s", mongodTmpFile.Name(), mongodPath, renameErr)
 	}
 

--- a/mongobin/getOrDownload_test.go
+++ b/mongobin/getOrDownload_test.go
@@ -46,3 +46,42 @@ func TestGetOrDownload(t *testing.T) {
 
 	assert.Equal(t, stat.ModTime(), stat2.ModTime())
 }
+
+func TestGetOrDownloadDifferentFilesystems(t *testing.T) {
+	mongobin.Afs = afero.Afero{Fs: afero.NewMemMapFs()}
+
+	spec := mongobin.DownloadSpec{
+		Version:        "4.0.5",
+		Platform:       "osx",
+		SSLBuildNeeded: true,
+		Arch:           "x86_64",
+	}
+
+	// Initialize the cache in a different filesystem
+	cacheFs := afero.Afero{Fs: afero.NewMemMapFs()}
+	cacheDir, err := cacheFs.TempDir("", "")
+	require.NoError(t, err)
+
+	// First call should download the file
+	path, err := mongobin.GetOrDownloadMongod(spec.GetDownloadURL(), cacheDir, memongolog.New(nil, memongolog.LogLevelDebug))
+	require.NoError(t, err)
+
+	assert.Equal(t, cacheDir+"/mongodb-osx-ssl-x86_64-4_0_5_tgz_d50ef2155b/mongod", path)
+
+	stat, err := mongobin.Afs.Stat(path)
+	require.NoError(t, err)
+
+	assert.True(t, stat.Size() > 50000000)
+	assert.True(t, stat.Mode()&0100 != 0)
+
+	// Second call should used the cached file
+	path2, err := mongobin.GetOrDownloadMongod(spec.GetDownloadURL(), cacheDir, memongolog.New(nil, memongolog.LogLevelDebug))
+	require.NoError(t, err)
+
+	assert.Equal(t, path, path2)
+
+	stat2, err := mongobin.Afs.Stat(path2)
+	require.NoError(t, err)
+
+	assert.Equal(t, stat.ModTime(), stat2.ModTime())
+}

